lib/Dialect/FIRRTL/Import/FIRParser.cpp

```diff
@@ -1055,18 +1055,40 @@ struct FIRStmtParser : public FIRScopedParser {
   /// a passive-typed value and return that.
   Value convertToPassive(Value input, Location loc);
 
+  /// Attach invalid values to every element of the value.
+  void emitInvalidate(Value val, Location loc, bool isDuplex) {
+    // Invalidate doesn't match the semantics of connect or partial connect, and
+    // so we have to manually do something like expand-connects. Invalid value
+    // needs to be connected to every field of a bundle which is flip, or
+    // every field in a duplex operation.
+    auto invalidType = val.getType().cast<FIRRTLType>();
+    if (invalidType.isa<AnalogType>()) {
+      // Analog types must be 'attach'ed, not connected.
+      auto invalidVal =
+          builder.create<InvalidValuePrimOp>(loc, invalidType.getPassiveType());
+      builder.create<AttachOp>(loc, ValueRange{val, invalidVal});
+    } else if (auto bundleType = invalidType.dyn_cast<BundleType>()) {
+      // We need to recurse into bundle types without an outer flip.  They could
+      // have inner flips or analog types.
+      for (auto element : bundleType.getElements()) {
+        auto elementVal = builder.create<SubfieldOp>(loc, val, element.name);
+        emitInvalidate(elementVal, loc, isDuplex);
+      }
+    } else if (invalidType.isa<FlipType>() || isDuplex) {
```

Comment:
Will this handle `flip<bundle<...>>` correctly? I think that formulation is possible if the internal bundle is passive. Or is the idea that this will then get picked up by lower types properly?

Reply:
Yeah, the idea is that this will create a bundle of invalids and this gets handled in lower-types. There is this test case for it:
```
    output out1 : { a : UInt<1>, b : UInt<1> }
    ; CHECK: [[INV:%.+]] = firrtl.invalidvalue
    ; CHECK: firrtl.connect %out1, [[INV]]
    out1 is invalid
 ```

Reply:
:100:

Reply:
This only works when the bundle has an outer flip.  There is an inefficiency here that I was too lazy/ it was too ugly to fix: It doesn't do the same sort of thing for a wire with non-flipped fields, it will generate a separate subfield+connect for each field :(